File: app/models/schedule.py
"""
Schedule models for monitoring automation
"""
from datetime import datetime
from typing import Optional, Dict, Any, Literal
from uuid import UUID
from pydantic import BaseModel, Field, validator
from enum import Enum
import json


class ScheduleType(str, Enum):
    """Schedule type enumeration"""
    INTERVAL = "interval"
    CRON = "cron"


class TaskType(str, Enum):
    """Task type enumeration"""
    CATEGORY_MONITOR = "category_monitor"
    PRODUCT_SCRAPE = "product_scrape"
    PRICE_UPDATE = "price_update"
    INVENTORY_CHECK = "inventory_check"


class ScheduleStatus(str, Enum):
    """Schedule execution status"""
    SUCCESS = "success"
    FAILURE = "failure"
    RUNNING = "running"


class IntervalSchedule(BaseModel):
    """Interval-based schedule configuration"""
    days: Optional[int] = 0
    hours: Optional[int] = 0
    minutes: Optional[int] = 0
    seconds: Optional[int] = 0
    
    @validator('*', pre=True)
    def validate_positive(cls, v):
        if v is not None and v < 0:
            raise ValueError("Interval values must be positive")
        return v
    
    def to_celery_schedule(self):
        """Convert to Celery schedule format"""
        from datetime import timedelta
        return timedelta(
            days=self.days or 0,
            hours=self.hours or 0,
            minutes=self.minutes or 0,
            seconds=self.seconds or 0
        )


class CronSchedule(BaseModel):
    """Cron-based schedule configuration"""
    minute: str = "*"
    hour: str = "*"
    day_of_week: str = "*"
    day_of_month: str = "*"
    month_of_year: str = "*"
    
    def to_celery_schedule(self):
        """Convert to Celery crontab format"""
        from celery.schedules import crontab
        return crontab(
            minute=self.minute,
            hour=self.hour,
            day_of_week=self.day_of_week,
            day_of_month=self.day_of_month,
            month_of_year=self.month_of_year
        )


class MonitoringSchedule(BaseModel):
    """Monitoring schedule model"""
    id: Optional[UUID] = None
    task_name: str = Field(..., max_length=100)
    task_type: TaskType = TaskType.CATEGORY_MONITOR
    description: Optional[str] = None
    
    # Schedule configuration
    schedule_type: ScheduleType
    schedule_value: Dict[str, Any]
    
    # Task parameters
    task_params: Dict[str, Any] = Field(default_factory=dict)
    
    # Status fields
    enabled: bool = True
    last_run: Optional[datetime] = None
    next_run: Optional[datetime] = None
    last_status: Optional[ScheduleStatus] = None
    last_error: Optional[str] = None
    
    # Metadata
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat() if v else None,
            UUID: lambda v: str(v) if v else None
        }
    
    def get_schedule_object(self):
        """Get the appropriate schedule object based on schedule_type"""
        schedule_type_value = self.schedule_type.value if hasattr(self.schedule_type, 'value') else self.schedule_type
        if schedule_type_value == ScheduleType.INTERVAL.value:
            return IntervalSchedule(**self.schedule_value)
        elif schedule_type_value == ScheduleType.CRON.value:
            return CronSchedule(**self.schedule_value)
        else:
            raise ValueError(f"Unknown schedule type: {self.schedule_type}")
    
    def to_celery_beat_schedule(self):
        """Convert to Celery Beat schedule entry"""
        schedule_obj = self.get_schedule_object()
        
        # Map task types to actual Celery tasks
        task_mapping = {
            TaskType.CATEGORY_MONITOR.value: "app.tasks.category_tasks.monitor_categories",
            TaskType.PRODUCT_SCRAPE.value: "app.tasks.scraping_tasks.scrape_products",
            TaskType.PRICE_UPDATE.value: "app.tasks.scraping_tasks.update_prices",
            TaskType.INVENTORY_CHECK.value: "app.tasks.scraping_tasks.check_inventory"
        }
        
        return {
            "task": task_mapping.get(self.task_type.value if hasattr(self.task_type, 'value') else self.task_type, "app.tasks.category_tasks.monitor_categories"),
            "schedule": schedule_obj.to_celery_schedule(),
            "args": [],
            "kwargs": self.task_params,
            "options": {
                "expires": 3600,  # Task expires after 1 hour if not executed
                "retry": True,
                "retry_policy": {
                    "max_retries": 3,
                    "interval_start": 60,
                    "interval_step": 120,
                    "interval_max": 600
                }
            }
        }


class ScheduleHistory(BaseModel):
    """Schedule execution history"""
    id: Optional[UUID] = None
    schedule_id: UUID
    started_at: datetime = Field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None
    status: ScheduleStatus
    result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    
    # Performance metrics
    duration_seconds: Optional[int] = None
    items_processed: Optional[int] = None
    items_failed: Optional[int] = None
    
    created_at: Optional[datetime] = None
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat() if v else None,
            UUID: lambda v: str(v) if v else None
        }


# Request/Response models for API
class ScheduleCreate(BaseModel):
    """Create schedule request"""
    task_name: str = Field(..., max_length=100)
    task_type: TaskType = TaskType.CATEGORY_MONITOR
    description: Optional[str] = None
    schedule_type: ScheduleType
    schedule_value: Dict[str, Any]
    task_params: Dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class ScheduleUpdate(BaseModel):
    """Update schedule request"""
    description: Optional[str] = None
    schedule_type: Optional[ScheduleType] = None
    schedule_value: Optional[Dict[str, Any]] = None
    task_params: Optional[Dict[str, Any]] = None
    enabled: Optional[bool] = None


class ScheduleResponse(MonitoringSchedule):
    """Schedule response with computed fields"""
    is_overdue: bool = False
    run_count: int = 0
    success_rate: float = 0.0
    
    @classmethod
    def from_db(cls, schedule_dict: Dict[str, Any], history_stats: Optional[Dict[str, Any]] = None):
        """Create response from database record with computed fields"""
        schedule = cls(**schedule_dict)
        
        if history_stats:
            schedule.run_count = history_stats.get('total_runs', 0)
            if schedule.run_count > 0:
                schedule.success_rate = history_stats.get('successful_runs', 0) / schedule.run_count * 100
        
        # Check if schedule is overdue
        if schedule.enabled and schedule.next_run:
            schedule.is_overdue = datetime.now() > schedule.next_run
        
        return schedule

File: frontend/src/types/schedule.ts
/**
 * Schedule types for monitoring automation
 */

export enum ScheduleType {
  INTERVAL = 'interval',
  CRON = 'cron'
}

export enum TaskType {
  CATEGORY_MONITOR = 'category_monitor',
  PRODUCT_SCRAPE = 'product_scrape',
  PRICE_UPDATE = 'price_update',
  INVENTORY_CHECK = 'inventory_check'
}

export enum ScheduleStatus {
  SUCCESS = 'success',
  FAILURE = 'failure',
  RUNNING = 'running'
}

export interface IntervalSchedule {
  days?: number;
  hours?: number;
  minutes?: number;
  seconds?: number;
}

export interface CronSchedule {
  minute: string;
  hour: string;
  day_of_week: string;
  day_of_month: string;
  month_of_year: string;
}

export interface MonitoringSchedule {
  id?: string;
  task_name: string;
  task_type: TaskType;
  description?: string;
  schedule_type: ScheduleType;
  schedule_value: IntervalSchedule | CronSchedule | Record<string, any>;
  task_params: Record<string, any>;
  enabled: boolean;
  last_run?: string;
  next_run?: string;
  last_status?: ScheduleStatus;
  last_error?: string;
  created_at?: string;
  updated_at?: string;
  created_by?: string;
  updated_by?: string;
}

export interface ScheduleResponse extends MonitoringSchedule {
  is_overdue: boolean;
  run_count: number;
  success_rate: number;
}

export interface ScheduleHistory {
  id: string;
  schedule_id: string;
  started_at: string;
  completed_at?: string;
  status: ScheduleStatus;
  result?: Record<string, any>;
  error_message?: string;
  duration_seconds?: number;
  items_processed?: number;
  items_failed?: number;
  created_at: string;
}

export interface ScheduleCreate {
  task_name: string;
  task_type: TaskType;
  description?: string;
  schedule_type: ScheduleType;
  schedule_value: IntervalSchedule | CronSchedule | Record<string, any>;
  task_params?: Record<string, any>;
  enabled?: boolean;
}

export interface ScheduleUpdate {
  description?: string;
  schedule_type?: ScheduleType;
  schedule_value?: IntervalSchedule | CronSchedule | Record<string, any>;
  task_params?: Record<string, any>;
  enabled?: boolean;
}

export interface ScheduleStats {
  total_schedules: number;
  enabled_schedules: number;
  disabled_schedules: number;
  schedules_by_type: Record<TaskType, number>;
  recent_runs: {
    total: number;
    successful: number;
    failed: number;
    running: number;
  };
  average_duration_seconds: number;
  next_scheduled_runs: Array<{
    task_name: string;
    task_type: TaskType;
    next_run: string;
  }>;
}

// Helper functions
export const getTaskTypeLabel = (taskType: TaskType): string => {
  switch (taskType) {
    case TaskType.CATEGORY_MONITOR:
      return 'Category Monitoring';
    case TaskType.PRODUCT_SCRAPE:
      return 'Product Scraping';
    case TaskType.PRICE_UPDATE:
      return 'Price Update';
    case TaskType.INVENTORY_CHECK:
      return 'Inventory Check';
    default:
      return taskType;
  }
};

export const getScheduleTypeLabel = (scheduleType: ScheduleType): string => {
  switch (scheduleType) {
    case ScheduleType.INTERVAL:
      return 'Interval';
    case ScheduleType.CRON:
      return 'Cron Expression';
    default:
      return scheduleType;
  }
};

export const formatInterval = (interval: IntervalSchedule): string => {
  const parts = [];
  if (interval.days) parts.push(`${interval.days} day${interval.days > 1 ? 's' : ''}`);
  if (interval.hours) parts.push(`${interval.hours} hour${interval.hours > 1 ? 's' : ''}`);
  if (interval.minutes) parts.push(`${interval.minutes} minute${interval.minutes > 1 ? 's' : ''}`);
  if (interval.seconds) parts.push(`${interval.seconds} second${interval.seconds > 1 ? 's' : ''}`);
  return parts.join(', ') || 'No interval';
};

export const formatCron = (cron: CronSchedule): string => {
  return `${cron.minute} ${cron.hour} ${cron.day_of_month} ${cron.month_of_year} ${cron.day_of_week}`;
};

export const getStatusColor = (status: ScheduleStatus): 'success' | 'error' | 'warning' | 'info' => {
  switch (status) {
    case ScheduleStatus.SUCCESS:
      return 'success';
    case ScheduleStatus.FAILURE:
      return 'error';
    case ScheduleStatus.RUNNING:
      return 'info';
    default:
      return 'warning';
  }
};